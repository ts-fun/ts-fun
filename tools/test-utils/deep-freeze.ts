function actualDeepFreeze<Value> (object: Value, visited: any[]): Value {
  Object.freeze(object)

  if (
    !object ||
    (typeof object !== 'object' && typeof object !== 'function') ||
    visited.includes(object)
  ) return object

  for (const value of Object.values(object)) {
    actualDeepFreeze(value, [...visited, object])
  }

  return object
}

export const deepFreeze = <Value> (object: Value) => actualDeepFreeze(object, [])
