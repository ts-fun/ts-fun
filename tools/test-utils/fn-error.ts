import { dbg } from 'string-template-format-inspect'

class NoError<Value> extends Error {
  constructor (
    public readonly fn: () => Value,
    public readonly value: Value
  ) {
    super(dbg`Expecting execution of ${fn} to throws an error but it returns ${value}`)
  }
}

export function fnError (fn: () => any) {
  let result: any

  try {
    result = fn()
  } catch (error) {
    return error
  }

  throw new NoError(fn, result)
}

export default fnError
