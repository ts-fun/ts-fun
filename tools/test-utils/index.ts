export * from './deep-freeze'
export * from './fn-error'
