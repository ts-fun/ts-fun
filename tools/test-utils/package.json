{
  "name": "@tools/test-utils",
  "version": "0.0.0",
  "private": true,
  "dependencies": {
    "string-template-format-inspect": "^0.1.4"
  }
}
