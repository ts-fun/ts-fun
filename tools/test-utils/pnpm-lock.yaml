dependencies:
  string-template-format-inspect: 0.1.8
lockfileVersion: 5.1
packages:
  /@types/node/14.6.1:
    dev: false
    resolution:
      integrity: sha512-HnYlg/BRF8uC1FyKRFZwRaCPTPYKa+6I8QiUZFLredaGOou481cgFS4wKRFyKvQtX8xudqkSdBczJHIYSQYKrQ==
  /string-template-format-base/0.1.7:
    dependencies:
      '@types/node': 14.6.1
      tslib: 2.0.1
    dev: false
    resolution:
      integrity: sha512-ySUoRCuRa6nhwJ6Xcu41VEpvwuDDVVPPqOnEf4P0UG4L860xbvC66YH0xno7dKW2JjN8jX0YstAcGk2q6Z6JzA==
  /string-template-format-inspect/0.1.8:
    dependencies:
      '@types/node': 14.6.1
      string-template-format-base: 0.1.7
      tslib: 2.0.1
    dev: false
    resolution:
      integrity: sha512-k3lQNBDZ8o6Q0o9WYXLUgKwDtmn1OpERFJZb4NPwG2MvgDD2iI6+9yDVK66WsyGI90ljF2wh/5KjO0E5tIE1Ig==
  /tslib/2.0.1:
    dev: false
    resolution:
      integrity: sha512-SgIkNheinmEBgx1IUNirK0TUD4X9yjjBRTqqjggWCU3pUEqIk3/Uwl3yRixYKT6WjQuGiwDv4NomL3wqRCj+CQ==
specifiers:
  string-template-format-inspect: ^0.1.8
