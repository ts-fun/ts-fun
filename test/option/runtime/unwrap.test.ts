import { fnError } from '@tools/test-utils'

import {
  unwrap,
  unwrapOrErr,
  unwrapOr,
  unwrapOrElse,
  some,
  none,
} from '@tsfun/option'

describe('unwrap', () => {
  describe('with a Some', () => {
    it('returns contained value', () => {
      expect(unwrap(some('value'))).toBe('value')
    })
  })

  describe('with a None', () => {
    it('throws provided error', () => {
      expect(fnError(() => unwrap(none()))).toMatchSnapshot()
    })
  })
})

describe('unwrapOrErr', () => {
  describe('with a Some', () => {
    it('returns contained value', () => {
      expect(unwrapOrErr(some('value'), 'error')).toBe('value')
    })
  })

  describe('with a None', () => {
    it('throws provided error', () => {
      expect(fnError(() => unwrapOrErr(none(), 'error'))).toBe('error')
    })
  })
})

describe('unwrapOr', () => {
  describe('with a Some', () => {
    it('returns contained value', () => {
      expect(unwrapOr(some('a'), 'b')).toBe('a')
    })
  })

  describe('with a None', () => {
    it('returns provided default', () => {
      expect(unwrapOr(none(), 'b')).toBe('b')
    })
  })
})

describe('unwrapOrElse', () => {
  const mkfn = () => jest.fn(() => 'default')

  describe('with a Some', () => {
    const fn = mkfn()
    const res = unwrapOrElse(some('value'), fn)

    it('returns contained value', () => {
      expect(res).toBe('value')
    })

    it('does not call provided function', () => {
      expect(fn).not.toBeCalled()
    })
  })

  describe('with a None', () => {
    const fn = mkfn()
    const res = unwrapOrElse(none(), fn)

    it('returns result of the function', () => {
      expect(res).toBe('default')
    })

    it('calls provided function exactly once with no arguments', () => {
      expect(fn.mock.calls).toEqual([[]])
    })
  })
})
