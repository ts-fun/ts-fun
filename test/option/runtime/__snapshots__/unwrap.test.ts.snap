// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`unwrap with a None throws provided error 1`] = `"Cannot unwrap a None"`;
